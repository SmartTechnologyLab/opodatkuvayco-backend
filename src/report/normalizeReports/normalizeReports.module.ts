import { Module } from '@nestjs/common';
import { NormalizeReportsService } from './normalizeReports.service';
import { NormalizeTradesService } from 'src/report/normalizeTrades/normalizeTrades.service';

@Module({
  providers: [NormalizeReportsService, NormalizeTradesService],
})
export class NormalizeReportsModule {}
