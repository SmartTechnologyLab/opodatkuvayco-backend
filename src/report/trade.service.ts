import { DealsService } from 'src/deals/deals.service';
import { Deal } from './types/interfaces/deal.interface';
import { Trade } from './types/interfaces/trade.interface';

export class DealService {
  deals: Deal[] = [];
  private buyQueue: Trade[] = [];
  private trades: Record<string, Trade[]> = {};
  private shortPositions: Trade[] = [];

  constructor(
    private dealService: DealsService,
    trades: Record<string, Trade[]>,
  ) {
    this.trades = trades;
  }

  async setDeal(buy: Trade, sell: Trade, commission: number) {
    const deal = await this.dealService.setDeal(buy, sell, commission);

    this.deals.push(deal);
  }

  async processTrades() {
    for (const trades in this.trades) {
      for (const trade of this.trades[trades]) {
        await this.processTrade(trade);
      }
    }
  }

  async processTrade(trade: Trade) {
    if (trade.operation === 'buy') {
      await this.processBuy(trade);
    } else if (trade.operation === 'sell') {
      if (this.buyQueue.length > 0) {
        await this.processLongSell(trade);
      } else {
        this.processShortSell(trade);
      }
    }
  }

  processShortSell(trade: Trade) {
    this.shortPositions.push(trade);
  }

  async processLongSell(sellTrade: Trade) {
    while (sellTrade.quantity > 0 && this.buyQueue.length > 0) {
      const buyTrade = this.buyQueue[0];

      const quantityToSell = Math.min(sellTrade.quantity, buyTrade.quantity);
      buyTrade.quantity -= quantityToSell;
      sellTrade.quantity -= quantityToSell;

      await this.setDeal(
        { ...buyTrade, quantity: quantityToSell },
        { ...sellTrade, quantity: quantityToSell },
        0,
      );

      if (buyTrade.quantity === 0) {
        this.buyQueue.shift();
      }
    }
  }

  async processBuy(buyTrade: Trade) {
    if (this.shortPositions.length > 0) {
      let remainingQuantity = buyTrade.quantity;

      while (remainingQuantity > 0 && this.shortPositions.length > 0) {
        const shortTrade = this.shortPositions[0];

        const quantityToCover = Math.min(
          remainingQuantity,
          shortTrade.quantity,
        );
        shortTrade.quantity -= quantityToCover;
        remainingQuantity -= quantityToCover;

        await this.setDeal(
          { ...shortTrade, quantity: quantityToCover },
          { ...buyTrade, quantity: quantityToCover },
          0,
        );

        if (!shortTrade.quantity) {
          this.shortPositions.shift();
        }
      }

      if (remainingQuantity) {
        this.buyQueue.push({ ...buyTrade, quantity: remainingQuantity });
      }
    } else {
      this.buyQueue.push(buyTrade);
    }
  }
}
