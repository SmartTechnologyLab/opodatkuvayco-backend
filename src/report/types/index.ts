export type OperationType = 'buy' | 'sell';

export interface IDealReport<T> {
  total: number;
  totalTaxFee: number;
  totalMilitaryFee: number;
  deals: T[];
}

export type GrouppedTrades = { [key: string]: ITrade[] };

export interface IReport<T> {
  dateStart: string;
  trades: T[];
}

export interface ITrade {
  date: string;
  price: number;
  commission: number;
  operation: OperationType;
  quantity: number;
  ticker: string;
  currency: string;
}

export type DealOptions = {
  purchaseCommission?: number;
  purchaseDate?: Date;
  purchasePrice?: number;
  purchaseRate?: number;
  quantity?: number;
  saleCommission?: number;
  saleDate?: Date;
  salePrice?: number;
  saleRate?: number;
  ticker?: string;
};

export type Deal = {
  id: symbol;
  percent: number;
  purchase: {
    date: Date;
    price: number;
    sum: number;
    commission: number;
    rate: number;
    uah: number;
  };
  quantity: number;
  sale: {
    date: Date;
    price: number;
    sum: number;
    commission: number;
    rate: number;
    uah: number;
  };
  ticker: string;
  total: number;
};
