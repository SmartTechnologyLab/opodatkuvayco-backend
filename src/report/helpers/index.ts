import { sort } from 'ramda';
import { IReport, ITrade } from '../types';

export const sortByDate = (report: IReport<ITrade>[]): IReport<ITrade>[] => {
  const byDate = (a: IReport<ITrade>, b: IReport<ITrade>) =>
    new Date(b.dateStart).getTime() - new Date(a.dateStart).getTime();
  return sort(byDate, report);
};
