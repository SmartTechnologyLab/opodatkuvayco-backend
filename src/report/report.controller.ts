import {
  Controller,
  Post,
  Query,
  UploadedFiles,
  UseInterceptors,
} from '@nestjs/common';
import { ReportService } from './report.service';
import { FilesInterceptor } from '@nestjs/platform-express';
import { Deal, IDealReport } from './types';
import { StockExchange } from '../normalizeTrades/constants';
import {
  ApiBody,
  ApiOperation,
  ApiQuery,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';

@ApiTags('report')
@Controller('report')
export class ReportController {
  constructor(private reportService: ReportService) {}

  @Post('deals')
  @UseInterceptors(FilesInterceptor('file', 10))
  @ApiOperation({ summary: 'Getting deals' })
  @ApiQuery({ name: 'type', description: 'Report type', required: false })
  @ApiBody({
    description: 'Loading trades report in JSON format for getting deals',
    required: true,
    schema: {
      type: 'object',
      properties: {
        file: {
          type: 'string',
          format: 'binary',
          description: 'JSON file containing trades report',
        },
      },
    },
  })
  @ApiResponse({
    status: 200,
    description: 'Returns deals',
  })
  async getDeals(
    @Query('type') type: string,
    @UploadedFiles() files: Express.Multer.File[],
  ): Promise<IDealReport<Deal>> {
    try {
      return this.reportService.handleReports(
        files,
        type,
        StockExchange.FREEDOM_FINANCE,
      );
    } catch (error) {
      throw new Error(error);
    }
  }
}
