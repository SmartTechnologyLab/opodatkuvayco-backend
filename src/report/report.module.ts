import { Module } from '@nestjs/common';
import { ReportService } from './report.service';
import { ReportController } from './report.controller';
import { CurrencyExchangeModule } from 'src/currencyExchange/currencyExchange.module';
import { CurrencyExchangeService } from 'src/currencyExchange/currencyExchange.service';
import { NormalizeTradesModule } from 'src/normalizeTrades/normalizeTrades.module';
import { NormalizeTradesService } from 'src/normalizeTrades/normalizeTrades.service';
import { NormalizeReportsModule } from 'src/normalizeReports/normalizeReports.module';
import { NormalizeReportsService } from 'src/normalizeReports/normalizeReports.service';

@Module({
  imports: [
    CurrencyExchangeModule,
    NormalizeTradesModule,
    NormalizeReportsModule,
  ],
  controllers: [ReportController],
  providers: [
    ReportService,
    CurrencyExchangeService,
    NormalizeTradesService,
    NormalizeReportsService,
  ],
})
export class ReportModule {}
