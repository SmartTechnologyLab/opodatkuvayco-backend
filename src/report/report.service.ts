import { BadRequestException, Injectable } from '@nestjs/common';
import { clone, reverse } from 'ramda';
import { CurrencyExchangeService } from '../currencyExchange/currencyExchange.service';
import { Deal, DealOptions, ITrade, IDealReport, IReport } from './types';
import { sortByDate } from './helpers';
import { NormalizeTradesService } from '../normalizeTrades/normalizeTrades.service';
import { StockExchange } from '../normalizeTrades/constants';
import { IFreedomFinanceCorporateAction } from './types/freedomFinance';
import { NormalizeReportsService } from '../normalizeReports/normalizeReports.service';

@Injectable()
export class ReportService {
  constructor(
    private currencyExchangeService: CurrencyExchangeService,
    private normalizeTradeService: NormalizeTradesService,
    private normalizeReportsService: NormalizeReportsService,
  ) {}

  readReport(file: Express.Multer.File) {
    try {
      const reportContent = file.buffer.toString('utf-8');
      return JSON.parse(reportContent);
    } catch (error) {
      throw new Error(error);
    }
  }

  async getReportExtended<T>(
    trades: T[],
    stockExhange: StockExchange,
  ): Promise<IDealReport<Deal>> {
    const deals: Deal[] = [];

    const groupedTrades = this.normalizeTradeService.groupTradesByTicker(
      clone(trades),
      stockExhange,
    );

    for (const ticker in groupedTrades) {
      let buyQueue: ITrade[] = [];
      let sellComission = 0;
      for (const deal of groupedTrades[ticker]) {
        if (deal.operation === 'buy' && deal.quantity > 0) {
          const existingDeal = this.findDealByDateAndPrice(buyQueue, deal);
          if (existingDeal) {
            existingDeal.quantity += deal.quantity;
            existingDeal.commission += deal.commission;
          } else {
            buyQueue.push(deal);
          }
        } else if (deal.operation === 'sell' && buyQueue.length > 0) {
          sellComission = deal.commission / deal.quantity;

          for (const purchaseDeal of buyQueue) {
            if (deal.quantity > 0 && purchaseDeal.quantity > 0) {
              const newDeal = await this.setDeal(
                purchaseDeal,
                deal,
                sellComission,
              );

              deals.push(newDeal);
            }
          }

          buyQueue = buyQueue.filter(
            (purchaseDeal) => purchaseDeal.quantity > 0,
          );

          while (
            deal.quantity > 0 &&
            (buyQueue.some((b) => b.quantity > 0) ||
              this.getShortBuy(groupedTrades[ticker], deal))
          ) {
            const foundShortBuy = buyQueue.some((b) => b.quantity > 0)
              ? buyQueue.filter((b) => b.quantity > 0).at(0)
              : this.getShortBuy(groupedTrades[ticker], deal);

            if (foundShortBuy && foundShortBuy.quantity > 0) {
              const newDeal = await this.setDeal(
                foundShortBuy,
                deal,
                sellComission,
              );

              deals.push(newDeal);
            }
          }
        } else if (
          deal.operation === 'sell' &&
          buyQueue.length === 0 &&
          this.getShortBuy(groupedTrades[ticker], deal)
        ) {
          sellComission = deal.commission / deal.quantity;

          const foundShortBuy = this.getShortBuy(groupedTrades[ticker], deal);

          if (foundShortBuy) {
            const newDeal = await this.setDeal(
              foundShortBuy,
              deal,
              sellComission,
            );

            deals.push(newDeal);
          }
        }
        sellComission = 0;
      }

      if (
        groupedTrades[ticker].some(
          (deal) => deal.operation === 'sell' && deal.quantity > 0,
        )
      ) {
        throw new BadRequestException('Not enough buy deals');
      }
    }

    const total = deals.reduce((acc, deal) => acc + deal.total, 0);

    const totalTaxFee = this.getTotalTaxFee(total);

    const totalMilitaryFee = this.getMilitaryFee(total);

    return {
      total,
      totalTaxFee,
      totalMilitaryFee,
      deals: deals.sort((a, b) => a.ticker.localeCompare(b.ticker)),
    };
  }

  async getPrevTrades<T>(
    trades: T[],
    stockExchange: StockExchange,
  ): Promise<ITrade[]> {
    const remainedPurchaseDeals: ITrade[] = [];

    const groupedTrades = this.normalizeTradeService.groupTradesByTicker(
      clone(trades),
      stockExchange,
    );

    for (const ticker in groupedTrades) {
      let buyQueue: ITrade[] = [];
      for (const deal of groupedTrades[ticker]) {
        if (deal.operation === 'buy' && deal.quantity > 0) {
          const existingDeal = this.findDealByDateAndPrice(buyQueue, deal);
          if (existingDeal) {
            existingDeal.quantity += deal.quantity;
            existingDeal.commission += deal.commission;
          } else {
            buyQueue.push(deal);
          }
        } else if (deal.operation === 'sell' && buyQueue.length > 0) {
          for (const purchaseDeal of buyQueue) {
            if (deal.quantity >= purchaseDeal.quantity) {
              const quantityToProcess = Math.min(
                purchaseDeal.quantity,
                deal.quantity,
              );
              deal.quantity -= quantityToProcess;
              purchaseDeal.quantity -= quantityToProcess;
            }
          }

          buyQueue = buyQueue.filter(
            (purchaseDeal) => purchaseDeal.quantity > 0,
          );

          while (
            deal.quantity > 0 &&
            (buyQueue.some((b) => b.quantity > 0) ||
              this.getShortBuy(groupedTrades[ticker], deal))
          ) {
            const foundShortBuy = buyQueue.some((b) => b.quantity > 0)
              ? buyQueue.filter((b) => b.quantity > 0).at(0)
              : this.getShortBuy(groupedTrades[ticker], deal);

            if (foundShortBuy && foundShortBuy.quantity > 0) {
              const quantityToProcess = Math.min(
                foundShortBuy.quantity,
                deal.quantity,
              );
              foundShortBuy.quantity -= quantityToProcess;
              deal.quantity -= quantityToProcess;
            }
          }
        } else if (
          deal.operation === 'sell' &&
          buyQueue.length === 0 &&
          this.getShortBuy(groupedTrades[ticker], deal)
        ) {
          const foundShortBuy = this.getShortBuy(groupedTrades[ticker], deal);

          if (foundShortBuy) {
            const quantityToProcess = Math.min(
              foundShortBuy.quantity,
              deal.quantity,
            );
            deal.quantity -= quantityToProcess;
            foundShortBuy.quantity -= quantityToProcess;
          }
        }
      }
      remainedPurchaseDeals.push(
        ...buyQueue.filter((deal) => deal.quantity > 0),
      );
    }

    return remainedPurchaseDeals;
  }

  async getReport(
    report: ITrade[],
    stockExchange: StockExchange,
  ): Promise<IDealReport<Deal>> {
    const trades = (await this.getReportExtended(
      report,
      stockExchange,
    )) as IDealReport<Deal>;

    const deals = Object.values(
      trades.deals.reduce(
        (acc, deal) => {
          if (!acc[deal.ticker]) {
            acc[deal.ticker] = deal;
          } else {
            acc[deal.ticker].sale.uah += deal.sale.uah;
            acc[deal.ticker].purchase.uah += deal.purchase.uah;
            acc[deal.ticker].total += deal.total;
          }
          return acc;
        },
        {} as Record<string, Deal>,
      ),
    );

    return {
      total: trades.total,
      totalTaxFee: trades.totalTaxFee,
      totalMilitaryFee: trades.totalMilitaryFee,
      deals,
    };
  }

  getShortBuy(trades: ITrade[], currentSellTrade: ITrade) {
    return trades.find(
      (trade, indx) =>
        trade.operation === 'buy' &&
        trade.quantity > 0 &&
        indx > trades.indexOf(currentSellTrade),
    );
  }

  async setDeal(
    purchaseDeal: ITrade,
    sellDeal: ITrade,
    sellComission?: number,
  ): Promise<Deal> {
    const [purchaseRate, saleRate] = await this.fetchPurchaseAndSellRate(
      purchaseDeal,
      sellDeal,
    );

    const quantityToProcess = Math.min(
      purchaseDeal.quantity,
      sellDeal.quantity,
    );

    const deal = this.getDeal({
      ticker: purchaseDeal.ticker,
      purchaseCommission: purchaseDeal.commission,
      purchaseDate: new Date(purchaseDeal.date),
      purchasePrice: purchaseDeal.price,
      purchaseRate,
      quantity: quantityToProcess,
      saleCommission: sellComission * purchaseDeal.quantity,
      saleDate: new Date(sellDeal.date),
      salePrice: sellDeal.price,
      saleRate,
    });

    sellDeal.quantity -= quantityToProcess;
    purchaseDeal.quantity -= quantityToProcess;

    return deal;
  }

  async handleReports(
    files: Express.Multer.File[],
    reportType: string,
    stockExchange: StockExchange,
  ): Promise<IDealReport<Deal>> {
    const getReportFunction =
      reportType === 'extended'
        ? this.getReportExtended.bind(this)
        : this.getReport.bind(this);

    if (files.length === 1) {
      const report = this.readReport(files.at(0));

      const { trades } = this.normalizeReportsService.getReportByStockExchange(
        report,
        stockExchange,
      );

      return getReportFunction(trades, stockExchange);
    }

    const reports: IReport<ITrade>[] = [];
    const dealsToCalculate: ITrade[] = [];

    files.forEach((file) => {
      const report = this.readReport(file);

      reports.push(
        this.normalizeReportsService.getReportByStockExchange(
          report,
          stockExchange,
        ) as IReport<ITrade>,
      );
    });

    const sortedReportsByDate: IReport<ITrade>[] = sortByDate(reports);

    const reversedReports: IReport<ITrade>[] = reverse(sortedReportsByDate);

    const remainedDealsMap = new Map<number, ITrade[]>();

    for (const [indx, statement] of Object.entries(reversedReports)) {
      const index = +indx;

      if (index !== sortedReportsByDate.length - 1) {
        const previousDeals = remainedDealsMap.get(index - 1) || [];

        const deals = await this.getPrevTrades(
          previousDeals.length
            ? [...previousDeals, ...statement.trades]
            : statement.trades,
          stockExchange,
        );

        if (!remainedDealsMap.has(index)) {
          remainedDealsMap.set(index, []);
        }

        if ((deals as ITrade[]).length) {
          remainedDealsMap.get(index)?.push(...deals);
        }

        if (index === sortedReportsByDate.length - 2 && deals.length) {
          dealsToCalculate.push(...deals);
        }
      }
    }

    return getReportFunction(
      [...dealsToCalculate, ...sortedReportsByDate.at(0).trades],
      stockExchange,
    );
  }

  async fetchPurchaseAndSellRate(purchaseDeal: ITrade, sellDeal: ITrade) {
    const [{ rate: purchaseRate }, { rate: sellRate }] = await Promise.all([
      this.currencyExchangeService.getCurrencyExchange(
        purchaseDeal.currency,
        purchaseDeal.date,
      ),

      this.currencyExchangeService.getCurrencyExchange(
        sellDeal.currency,
        sellDeal.date,
      ),
    ]);

    return [purchaseRate, sellRate];
  }

  findDealByDateAndPrice(deals: ITrade[], currentDeal: ITrade) {
    return deals.find(
      (deal) =>
        this.currencyExchangeService.formatDateForCurrencyExchange(
          deal.date,
        ) ===
          this.currencyExchangeService.formatDateForCurrencyExchange(
            currentDeal.date,
          ) && deal.price === currentDeal.price,
    );
  }

  getTotalTaxFee(total: number) {
    if (total <= 0) {
      return 0;
    }

    return total * 0.18;
  }

  getMilitaryFee(total: number) {
    if (total <= 0) {
      return 0;
    }

    return total * 0.015;
  }

  getDeal({
    ticker,
    quantity,
    purchaseCommission,
    purchaseDate,
    purchasePrice,
    purchaseRate,
    saleCommission,
    saleDate,
    salePrice,
    saleRate,
  }: DealOptions = {}) {
    const purchaseSum = purchasePrice * quantity;
    const _purchaseCommission = purchaseCommission || purchaseSum;
    const saleSum = salePrice * quantity;
    const _saleCommission = saleCommission || saleSum;
    const purchaseUah =
      (purchaseSum + _purchaseCommission) * purchaseRate +
      _saleCommission * saleRate;
    const saleUah = saleSum * saleRate;

    return {
      id: Symbol(),
      purchase: {
        commission: _purchaseCommission,
        date: purchaseDate,
        price: purchasePrice,
        rate: purchaseRate,
        sum: purchaseSum,
        uah: purchaseUah,
      },
      quantity: quantity,
      sale: {
        commission: _saleCommission,
        date: saleDate,
        price: salePrice,
        rate: saleRate,
        sum: saleSum,
        uah: saleUah,
      },
      ticker: ticker,
      total: saleUah - purchaseUah,
      percent: saleUah / purchaseUah - 1,
    };
  }

  async calculateDividends(file: Express.Multer.File) {
    const corporateActions = this.readReport(file).corporate_actions
      .detailed as IFreedomFinanceCorporateAction[];

    const filteredActionsByDividend = corporateActions.filter(
      (action) => action.type_id === 'dividend',
    );

    const dividends = await Promise.all(
      filteredActionsByDividend.map(async (dividend) => {
        const { rate } = await this.currencyExchangeService.getCurrencyExchange(
          dividend.currency,
          dividend.date,
        );

        return {
          ticker: dividend.ticker,
          rate,
          price: dividend.amount,
          uah: dividend.amount * rate,
        };
      }),
    );

    const totalDividends = dividends.reduce(
      (acc, dividend) => acc + dividend.uah,
      0,
    );

    return {
      total: {
        sumUAH: totalDividends,
        taxFee: totalDividends * 0.09,
        militaryFee: totalDividends * 0.015,
      },
      dividendsResult: dividends,
    };
  }
}
