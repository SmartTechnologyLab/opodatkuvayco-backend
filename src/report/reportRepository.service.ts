import { BadRequestException, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Report as ReportEntity } from 'src/report/entities/report.entity';
import { DealReport } from './types/interfaces/deal-report.interface';
import { Deal } from './types/interfaces/deal.interface';
import { DealsService } from 'src/deals/deals.service';
import { User } from 'src/user/entities/user.entity';

@Injectable()
export class ReportRepositoryService {
  constructor(
    @InjectRepository(ReportEntity)
    private reportRepository: Repository<ReportEntity>,
    private dealsService: DealsService,
  ) {}

  async getReports(userId: User['id']): Promise<ReportEntity[]> {
    const report = await this.reportRepository.find({
      where: {
        user: {
          id: userId,
        },
      },
      relations: ['deals', 'deals.purchase', 'deals.sale'],
    });

    return report;
  }

  async saveReport(
    report: DealReport<Deal>,
    user: User,
  ): Promise<ReportEntity> {
    try {
      const deals = await this.dealsService.saveDeals(report.deals, user);

      const newReport = new ReportEntity();

      newReport.total = report.total;
      newReport.totalMilitaryFee = report.totalMilitaryFee;
      newReport.totalTaxFee = report.totalTaxFee;
      newReport.deals = deals;
      newReport.user = user;

      return this.reportRepository.save(newReport);
    } catch (error) {
      throw new BadRequestException(error.message);
    }
  }
}
