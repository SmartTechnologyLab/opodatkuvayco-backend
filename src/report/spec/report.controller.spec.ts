import { CurrencyExchangeService } from '../../currencyExchange/currencyExchange.service';
import { ReportController } from '../report.controller';
import { ReportService } from '../report.service';
import * as request from 'supertest';
import { Test, TestingModule } from '@nestjs/testing';
import { trades } from './__fixtures__/report';
import { resultDeals, shortResultDeals } from './__fixtures__/resultDeals';
import { INestApplication } from '@nestjs/common';
import { NormalizeTradesService } from '../../normalizeTrades/normalizeTrades.service';
import { NormalizeReportsService } from '../../normalizeReports/normalizeReports.service';

jest.mock('../report.service', () => ({
  ReportService: jest.fn().mockImplementation(() => {
    return {
      handleReports: jest.fn(),
    };
  }),
}));

describe('Report Controller', () => {
  let reportService: ReportService;
  let reportController: ReportController;
  let app: INestApplication;

  beforeEach(async () => {
    const moduleRef: TestingModule = await Test.createTestingModule({
      controllers: [ReportController],
      providers: [
        ReportService,
        {
          provide: CurrencyExchangeService,
          useValue: new CurrencyExchangeService(),
        },
        {
          provide: NormalizeTradesService,
          useValue: new NormalizeTradesService(),
        },
        {
          provide: NormalizeReportsService,
          useValue: new NormalizeReportsService(),
        },
      ],
    }).compile();

    reportController = moduleRef.get<ReportController>(ReportController);
    reportService = moduleRef.get<ReportService>(ReportService);

    app = moduleRef.createNestApplication();
    await app.init();
  });

  afterEach(() => {
    jest.clearAllMocks();

    if (app) {
      app.close();
    }
  });

  it('report controller should be defined', () => {
    expect(reportController).toBeDefined();
  });

  describe('Post /deals', () => {
    it('should return extended report when query type is "extended"', async () => {
      const mockFile = {
        buffer: Buffer.from(JSON.stringify({ trades: { detailed: trades } })),
      } as Express.Multer.File;

      jest.spyOn(reportService, 'handleReports').mockResolvedValue(resultDeals);

      const response = await request(app.getHttpServer())
        .post('/report/deals')
        .query({ type: 'extended' })
        .attach('file', mockFile.buffer, 'report.json');

      expect(response.status).toBe(201);
      expect(response.body).toMatchSnapshot();
    });

    it('should return short report when query type is empty', async () => {
      const mockFile = {
        buffer: Buffer.from(JSON.stringify({ trades: { detailed: trades } })),
      } as Express.Multer.File;

      jest
        .spyOn(reportService, 'handleReports')
        .mockResolvedValue(shortResultDeals);

      const response = await request(app.getHttpServer())
        .post('/report/deals')
        .query({ type: '' })
        .attach('file', mockFile.buffer, 'report.json');

      expect(response.body).toMatchSnapshot();
    });
  });
});
