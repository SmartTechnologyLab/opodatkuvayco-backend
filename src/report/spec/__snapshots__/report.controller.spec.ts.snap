// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Controller Post /deals should return extended report when query type is "extended" 1`] = `
{
  "deals": [
    {
      "percent": -0.007445928377260325,
      "purchase": {
        "commission": 4.62,
        "date": "2021-06-07T08:48:21.000Z",
        "price": 35.53,
        "rate": 44,
        "sum": 923.78,
        "uah": 40951.24,
      },
      "quantity": 26,
      "sale": {
        "commission": 2.31,
        "date": "2021-06-07T08:48:21.000Z",
        "price": 35.53,
        "rate": 44,
        "sum": 923.78,
        "uah": 40646.32,
      },
      "ticker": "FIPO",
      "total": -304.91999999999825,
    },
    {
      "percent": -0.031465987631866144,
      "purchase": {
        "commission": 1.73,
        "date": "2021-06-11T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 4838.24,
      },
      "quantity": 10,
      "sale": {
        "commission": 1.73,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 4686,
      },
      "ticker": "GOOGL",
      "total": -152.23999999999978,
    },
    {
      "percent": -0.0625,
      "purchase": {
        "commission": 7.75,
        "date": "2021-06-10T11:06:17.000Z",
        "price": 155,
        "rate": 44,
        "sum": 155,
        "uah": 7274.666666666667,
      },
      "quantity": 1,
      "sale": {
        "commission": 2.5833333333333335,
        "date": "2021-10-10T11:06:17.000Z",
        "price": 155,
        "rate": 44,
        "sum": 155,
        "uah": 6820,
      },
      "ticker": "MNDY_IPO",
      "total": -454.66666666666697,
    },
    {
      "percent": -0.0382626680455016,
      "purchase": {
        "commission": 7.75,
        "date": "2021-07-10T11:06:17.000Z",
        "price": 156,
        "rate": 44,
        "sum": 468,
        "uah": 21274,
      },
      "quantity": 3,
      "sale": {
        "commission": 7.75,
        "date": "2021-10-10T11:06:17.000Z",
        "price": 155,
        "rate": 44,
        "sum": 465,
        "uah": 20460,
      },
      "ticker": "MNDY_IPO",
      "total": -814,
    },
    {
      "percent": -0.09620991253644318,
      "purchase": {
        "commission": 7.75,
        "date": "2021-07-10T11:06:17.000Z",
        "price": 156,
        "rate": 44,
        "sum": 156,
        "uah": 7546,
      },
      "quantity": 1,
      "sale": {
        "commission": 7.75,
        "date": "2021-11-10T12:06:17.000Z",
        "price": 155,
        "rate": 44,
        "sum": 155,
        "uah": 6820,
      },
      "ticker": "MNDY_IPO",
      "total": -726,
    },
    {
      "percent": -0.028919654119113125,
      "purchase": {
        "commission": 1.73,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 4825.553333333333,
      },
      "quantity": 10,
      "sale": {
        "commission": 1.4416666666666667,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 4686,
      },
      "ticker": "RIDE",
      "total": -139.55333333333328,
    },
    {
      "percent": -0.08655564291330131,
      "purchase": {
        "commission": 1.73,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 21.3,
        "uah": 1026.0066666666667,
      },
      "quantity": 2,
      "sale": {
        "commission": 0.28833333333333333,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 21.3,
        "uah": 937.2,
      },
      "ticker": "RIDE",
      "total": -88.80666666666662,
    },
    {
      "percent": -0.030674846625766916,
      "purchase": {
        "commission": 1.75,
        "date": "2021-06-10T13:30:00.000Z",
        "price": 11.06,
        "rate": 44,
        "sum": 110.60000000000001,
        "uah": 5020.400000000001,
      },
      "quantity": 10,
      "sale": {
        "commission": 1.75,
        "date": "2021-06-10T13:30:00.000Z",
        "price": 11.06,
        "rate": 44,
        "sum": 110.60000000000001,
        "uah": 4866.400000000001,
      },
      "ticker": "WISH",
      "total": -154,
    },
  ],
  "total": -2834.1866666666647,
  "totalMilitaryFee": 0,
  "totalTaxFee": 0,
}
`;

exports[`Report Controller Post /deals should return short report when query type is empty 1`] = `
{
  "deals": [
    {
      "percent": 0.05947617315387421,
      "purchase": {
        "commission": 1.73,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 4838.24,
      },
      "quantity": 10,
      "sale": {
        "commission": 1.73,
        "date": "2021-06-10T15:41:12.000Z",
        "price": 10.65,
        "rate": 44,
        "sum": 106.5,
        "uah": 5126,
      },
      "ticker": "APPL",
      "total": 287.7600000000002,
    },
  ],
  "total": 287.7600000000002,
  "totalMilitaryFee": 4.316400000000003,
  "totalTaxFee": 51.79680000000004,
}
`;
